reframe/core/systems.py

```diff
@@ -13,6 +13,187 @@
+    '''A representation of a processor inside ReFrame.
+
+    .. warning::
+       Users may not create :class:`Processor` objects directly.
+    '''
+
+    def __init__(self, processor_info):
+        self._arch = None
+        self._num_cpus = None
+        self._num_cpus_per_core = None
+        self._num_cpus_per_socket = None
+        self._num_sockets = None
+        self._topology = None
+        self._info = processor_info
+
+        if not processor_info:
+            return
+
+        for key, val in processor_info.items():
+            setattr(self, f'_{key}', val)
+
+    @property
+    def info(self):
+        '''All the available information from the configuration.
+
+        :type: :class:`dict`
+        '''
+        return self._info
+
+    @property
+    def arch(self):
+        '''The microarchitecture of the processors.
+
+        :type: :class:`str` or :class:`None`
+        '''
+        return self._arch
+
+    @property
+    def num_cpus(self):
+        '''Number of logical CPUs.
+
+        :type: integral or :class:`None`
+        '''
+        return self._num_cpus
+
+    @property
+    def num_cpus_per_core(self):
+        '''Number of logical CPUs per core.
+
+        :type: integral or :class:`None`
+        '''
+        return self._num_cpus_per_core
+
+    @property
+    def num_cpus_per_socket(self):
+        '''Number of logical CPUs per socket.
+
+        :type: integral or :class:`None`
+        '''
+        return self._num_cpus_per_socket
+
+    @property
+    def num_sockets(self):
+        '''Number of sockets.
+
+        :type: integral or :class:`None`
+        '''
+        return self._num_sockets
+
+    @property
+    def topology(self):
+        '''Processor topology.
+
+        :type: :class:`Dict[str, obj]` or :class:`None`
+        '''
+        return self._topology
+
+    @property
+    def num_cores(self):
+        '''Total number of cores.
+
+        :type: integral or :class:`None`
+        '''
+        if self._num_cpus and self._num_cpus_per_core:
+            return self._num_cpus // self._num_cpus_per_core
+        else:
+            return None
+
+    @property
+    def num_cores_per_socket(self):
+        '''Number of cores per socket.
+
+        :type: integral or :class:`None`
+        '''
+        if self.num_cores and self._num_sockets:
+            return self.num_cores // self._num_sockets
+        else:
+            return None
+
+    @property
+    def num_numa_nodes(self):
+        '''Number of numa nodes.
+
+        :type: integral or :class:`None`
+        '''
+        if self._topology and 'numa_nodes' in self._topology:
+            return len(self._topology['numa_nodes'])
+        else:
+            return None
+
+    @property
+    def num_cores_per_numa_node(self):
+        if self.num_numa_nodes and self.num_cores:
+            return self.num_cores // self.num_numa_nodes
+        else:
+            return None
+
+
+class Device(jsonext.JSONSerializable):
```

Comment:
How about changing this class name to `DeviceType` instead? If I understand this correctly, this class represents a type of device that is present in a partition, and this partition might have several devices of this type. So by calling this class `Device` I got the impression that this was referring instead to a single gpu and I found a bit confusing that the class `Device` had the attribute `_num_devices`. If we change this name to `DeviceType`, I'd suggest to also change the class above referring to the processor as `ProcessorType`, just for consistency purposes.

Reply:
Good point, I will change both `DeviceType` and `ProcessorType`.